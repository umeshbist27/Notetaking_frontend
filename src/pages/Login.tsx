import { toast } from "react-toastify";
import axios from "axios";
import { useState } from "react";
import React from "react";
import { useNavigate } from "react-router-dom";
import { loginSchema } from "../validation/authValidation";
import { Formik, ErrorMessage, Field, Form } from "formik";
import { Eye, EyeOff } from "lucide-react";
import { VITE_API } from "../config";

interface formValue{
  email:String,
  password:String
}

const Login = () => {
  const [showPassword, setShowPassword] = useState(false);
  const navigate = useNavigate();

  const handleSubmit = async (values:formValue, { resetForm, setSubmitting }: any) => {
    try {
      const res = await axios.post(
        `${VITE_API}api/auth/login`,
        values
      );
      localStorage.setItem("token", res.data.token);

      localStorage.setItem("username", res.data.name);

      resetForm();

      toast.success(`Logged in successfully`);
      navigate("/notes");
    } catch (err) {
      toast.error("Login failed. Please check your credentials.");
      console.error("Error occurred", err);
    } finally {
      setSubmitting(false);
    }
  };

  return (
    <div className="min-h-screen flex items-center justify-center bg-gray-100 px-4">
      <div className="w-full max-w-md bg-white border border-gray-200 rounded-xl shadow-lg p-8 space-y-6">
        <h1 className="text-3xl font-bold text-center text-gray-800">
          Welcome Back
        </h1>
        <p className="text-center text-gray-500">
          Please log in to your account
        </p>
        <Formik
          initialValues={{ email: "", password: "" }}
          onSubmit={handleSubmit}
          validationSchema={loginSchema}
          validateOnChange={true}
        >
          {({ isSubmitting }) => (
            <Form className="space-y-5">
              <div>
                <label
                  htmlFor="email"
                  className="block text-sm font-medium text-gray-700 mb-1"
                >
                  Email
                </label>
                <Field
                  type="email"
                  name="email"
                  autoComplete="current-email"
                  placeholder="Enter your email"
                  className="w-full px-4 py-2 border border-gray-300 rounded-lg focus:outline-none focus:ring-2 focus:ring-gray-400"
                />

                <ErrorMessage
                  name="email"
                  component="div"
                  className="text-red-500 text-xs mt-1"
                />
              </div>
              <div className="relative">
                <label
                  htmlFor="password"
                  className="block text-sm font-medium text-gray-700"
                ></label>
                <Field
                  name="password"
                  type={showPassword ? "text" : "password"}
                  autoComplete="current-password"
                  placeholder="********"
                  className="w-full px-4 py-2 border border-gray-300 rounded-lg focus:outline-none focus:ring-2 focus:ring-gray-400 pr-10"
                />
                <div
                  data-testid="toggle-password"
                  className="absolute inset-y-0 right-2 flex items-center cursor-pointer text-gray-500"
                  onClick={() => setShowPassword((prev) => !prev)}
                >
                  {showPassword ? <EyeOff size={18} /> : <Eye size={18} />}
                </div>
                <ErrorMessage
                  name="password"
                  component="div"
                  className="text-red-500 text-xs mt-1"
                />
              </div>
              <button
                type="submit"
                disabled={isSubmitting}
                className="w-full bg-gray-700 hover:bg-gray-800 text-white py-2 rounded-lg font-semibold transition duration-300"
              >
                Login
              </button>
            </Form>
          )}
        </Formik>

        <p className="text-center text-sm text-gray-600">
          Don't have an account?{" "}
          <span
            className="text-gray-700 font-medium cursor-pointer hover:underline"
            onClick={() => navigate("/signup")}
          >
            Sign up
          </span>
        </p>
      </div>
    </div>
  );
};

export default Login;
