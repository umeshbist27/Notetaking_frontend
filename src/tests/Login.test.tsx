import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";

import { BrowserRouter } from "react-router-dom";
import axios from "axios";
import { toast } from "react-toastify";
import Login from "../pages/Login";

jest.mock("axios");
jest.mock("react-toastify", () => ({
  toast: {
    success: jest.fn(),
    error: jest.fn(),
  },
}));

jest.mock('../config', () => ({
  VITE_API: 'http://mock-api:3000/api/auth/notes',
}));

const mockedNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedNavigate,
}));


const mockedAxios = axios as jest.Mocked<typeof axios>;

const renderWithRouter = (ui: React.ReactElement) => {
  return render(<BrowserRouter>{ui}</BrowserRouter>);
};

describe("Login Component", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    localStorage.clear();
  });

 it("toggles password visibility when eye icon is clicked", () => {
  renderWithRouter(<Login />);
  const passwordInput = screen.getByPlaceholderText("********") as HTMLInputElement;
  const toggleButton = screen.getByTestId("toggle-password");

  expect(passwordInput.type).toBe("password");

  fireEvent.click(toggleButton);
  expect(passwordInput.type).toBe("text");

  fireEvent.click(toggleButton);
  expect(passwordInput.type).toBe("password");
});

it("navigates to signup page when clicking 'Sign up'", () => {
  renderWithRouter(<Login />);
  const signupLink = screen.getByText("Sign up");
  fireEvent.click(signupLink);
  expect(mockedNavigate).toHaveBeenCalledWith("/signup");
});


  it("renders email and password fields", () => {
    renderWithRouter(<Login />);
    expect(screen.getByPlaceholderText("Enter your email")).toBeInTheDocument();
    expect(screen.getByPlaceholderText("********")).toBeInTheDocument();
    expect(screen.getByRole("button", { name: /login/i })).toBeInTheDocument();
  });

  it("logs in successfully and stores token", async () => {
    mockedAxios.post.mockResolvedValueOnce({
      data: {
        token: "fake-token",
        name: "Umesh",
      },
    });

    renderWithRouter(<Login />);

    fireEvent.change(screen.getByPlaceholderText("Enter your email"), {
      target: { value: "test@example.com" },
    });
    fireEvent.change(screen.getByPlaceholderText("********"), {
      target: { value: "password123" },
    });
    fireEvent.click(screen.getByRole("button", { name: /login/i }));

    await waitFor(() => {
      expect(mockedAxios.post).toHaveBeenCalledWith(
        "http://mock-api:3000/api/auth/notes/login",
        {
          email: "test@example.com",
          password: "password123",
        }
      );
      expect(localStorage.getItem("token")).toBe("fake-token");
      expect(localStorage.getItem("username")).toBe("Umesh");
      expect(toast.success).toHaveBeenCalledWith("Logged in successfully");
    });
  });

  it("shows error toast on login failure", async () => {
    mockedAxios.post.mockRejectedValueOnce(new Error("Invalid credentials"));

    renderWithRouter(<Login />);

    fireEvent.change(screen.getByPlaceholderText("Enter your email"), {
      target: { value: "wrong@example.com" },
    });
    fireEvent.change(screen.getByPlaceholderText("********"), {
      target: { value: "wrongpass" },
    });

    fireEvent.click(screen.getByRole("button", { name: /login/i }));

    await waitFor(() => {
      expect(toast.error).toHaveBeenCalledWith(
        "Login failed. Please check your credentials."
      );
    });
  });

  it("matches the snapshot", () => {
    const { asFragment } = renderWithRouter(<Login />);
    expect(asFragment()).toMatchSnapshot();
  });
});
