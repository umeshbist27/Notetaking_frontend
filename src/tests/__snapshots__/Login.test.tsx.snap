// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Login Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="min-h-screen flex items-center justify-center bg-gray-100 px-4"
  >
    <div
      class="w-full max-w-md bg-white border border-gray-200 rounded-xl shadow-lg p-8 space-y-6"
    >
      <h1
        class="text-3xl font-bold text-center text-gray-800"
      >
        Welcome Back
      </h1>
      <p
        class="text-center text-gray-500"
      >
        Please log in to your account
      </p>
      <form
        action="#"
        class="space-y-5"
      >
        <div>
          <label
            class="block text-sm font-medium text-gray-700 mb-1"
            for="email"
          >
            Email
          </label>
          <input
            autocomplete="current-email"
            class="w-full px-4 py-2 border border-gray-300 rounded-lg focus:outline-none focus:ring-2 focus:ring-gray-400"
            name="email"
            placeholder="Enter your email"
            type="email"
            value=""
          />
        </div>
        <div
          class="relative"
        >
          <label
            class="block text-sm font-medium text-gray-700"
            for="password"
          />
          <input
            autocomplete="current-password"
            class="w-full px-4 py-2 border border-gray-300 rounded-lg focus:outline-none focus:ring-2 focus:ring-gray-400 pr-10"
            name="password"
            placeholder="********"
            type="password"
            value=""
          />
          <div
            class="absolute inset-y-0 right-2 flex items-center cursor-pointer text-gray-500"
            data-testid="toggle-password"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-eye"
              fill="none"
              height="18"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="18"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M2.062 12.348a1 1 0 0 1 0-.696 10.75 10.75 0 0 1 19.876 0 1 1 0 0 1 0 .696 10.75 10.75 0 0 1-19.876 0"
              />
              <circle
                cx="12"
                cy="12"
                r="3"
              />
            </svg>
          </div>
        </div>
        <button
          class="w-full bg-gray-700 hover:bg-gray-800 text-white py-2 rounded-lg font-semibold transition duration-300"
          type="submit"
        >
          Login
        </button>
      </form>
      <p
        class="text-center text-sm text-gray-600"
      >
        Don't have an account? 
        <span
          class="text-gray-700 font-medium cursor-pointer hover:underline"
        >
          Sign up
        </span>
      </p>
    </div>
  </div>
</DocumentFragment>
`;
